### PR TITLE
Unit tests for Object creation and property getter/setter

### PR DESCRIPTION
**Pull request by a-ivanov**
More `Object` tests as was noted in #43440.

### CODE REVIEW

**Comment by Xrayez** [comment]
Debugging on Windows gives me crash with the following backtrace:
```
ntdll.dll!00007ffb602b8fe2() (Unknown Source:0)
ntdll.dll!00007ffb602c1412() (Unknown Source:0)
ntdll.dll!00007ffb602c171a() (Unknown Source:0)
ntdll.dll!00007ffb602ca6d9() (Unknown Source:0)
ntdll.dll!00007ffb6020135d() (Unknown Source:0)
ntdll.dll!00007ffb602006e1() (Unknown Source:0)
godot.windows.tools.64.exe!_free_base(void * block) Line 105 (c:\Users\Xrayez\AppData\Local\Programs\Microsoft VS Code\minkernel\crts\ucrt\src\appcrt\heap\free_base.cpp:105)
godot.windows.tools.64.exe!Memory::free_static(void * p_ptr, bool p_pad_align) Line 170 (core\os\memory.cpp:170)
godot.windows.tools.64.exe!memdelete<ScriptInstance>(ScriptInstance * p_class) Line 119 (core\os\memory.h:119)
godot.windows.tools.64.exe!Object::~Object() Line 1776 (core\object\object.cpp:1776)
godot.windows.tools.64.exe!TestObject::_DOCTEST_ANON_FUNC_1791() Line 219 (tests\test_object.h:219)
godot.windows.tools.64.exe!doctest::Context::run() Line 6168 (thirdparty\doctest\doctest.h:6168)
godot.windows.tools.64.exe!test_main(int argc, char * * argv) Line 130 (tests\test_main.cpp:130)
godot.windows.tools.64.exe!Main::test_entrypoint(int argc, char * * argv, bool & tests_need_run) Line 469 (main\main.cpp:469)
godot.windows.tools.64.exe!widechar_main(int argc, wchar_t * * argv) Line 149 (platform\windows\godot_windows.cpp:149)
godot.windows.tools.64.exe!_main() Line 185 (platform\windows\godot_windows.cpp:185)
godot.windows.tools.64.exe!main(int argc, char * * argv) Line 199 (platform\windows\godot_windows.cpp:199)
[Inline Frame] godot.windows.tools.64.exe!invoke_main() Line 78 (d:\agent\_work\63\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78)
godot.windows.tools.64.exe!__scrt_common_main_seh() Line 288 (d:\agent\_work\63\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
kernel32.dll!00007ffb60027c24() (Unknown Source:0)
ntdll.dll!00007ffb6022d4d1() (Unknown Source:0)
```

You'll also want to run the binary with `--test --no-breaks` so the debugger won't break on ClassDB tests currently.

File: tests/test_object.h [hunk 1]
```diff
@@ -87,6 +187,119 @@ TEST_CASE("[Object] Metadata") {
...
+	Object *p_db = ObjectDB::get_instance(object.get_instance_id());
+	CHECK_MESSAGE(
+			p_db == &object,
+			"The database pointer returned by the object id should reference same object.");
+}
+
+TEST_CASE("[Object] Script instance property setter") {
+	Object object;
+	_MockScriptInstance script_instance;
+	object.set_script_instance(&script_instance);
```
Comment (Xrayez) [inline tests/test_object.h:207]: This seems to resolve the crash above:

```suggestion
	_MockScriptInstance *script_instance = memnew(_MockScriptInstance);
	object.set_script_instance(script_instance);
```

Script instance will be freed automatically by the `object`'s destructor, which actually caused the crash.

Generally, you'll also likely want to instantiate `Object` with `memnew` as well for consistency with the rest of the codebase (just like `Node`s), but don't forget to `memdelete` those as well.

The same applies to other test cases.

File: tests/test_object.h [hunk 2]
```diff
@@ -87,6 +187,119 @@ TEST_CASE("[Object] Metadata") {
...
+
+	bool valid = false;
+	object.set("some_name", 100, &valid);
+	CHECK(valid);
+	Variant actual_value;
+	CHECK_MESSAGE(
+			script_instance.get("some_name", actual_value),
```
Comment (Xrayez) [inline tests/test_object.h:214]: Once the above is done:
```suggestion
			script_instance->get("some_name", actual_value),
```

**Review by Xrayez — CHANGES_REQUESTED** [review]

**Comment by a-ivanov** [comment]
Thanks @Xrayez! I will definitely fix these, a little busy ATM.

**Comment by a-ivanov** [comment]
> Debugging on Windows gives me crash with the following backtrace:
> 
> ```
> ntdll.dll!00007ffb602b8fe2() (Unknown Source:0)
> ntdll.dll!00007ffb602c1412() (Unknown Source:0)
> ntdll.dll!00007ffb602c171a() (Unknown Source:0)
> ntdll.dll!00007ffb602ca6d9() (Unknown Source:0)
> ntdll.dll!00007ffb6020135d() (Unknown Source:0)
> ntdll.dll!00007ffb602006e1() (Unknown Source:0)
> godot.windows.tools.64.exe!_free_base(void * block) Line 105 (c:\Users\Xrayez\AppData\Local\Programs\Microsoft VS Code\minkernel\crts\ucrt\src\appcrt\heap\free_base.cpp:105)
> godot.windows.tools.64.exe!Memory::free_static(void * p_ptr, bool p_pad_align) Line 170 (core\os\memory.cpp:170)
> godot.windows.tools.64.exe!memdelete<ScriptInstance>(ScriptInstance * p_class) Line 119 (core\os\memory.h:119)
> godot.windows.tools.64.exe!Object::~Object() Line 1776 (core\object\object.cpp:1776)
> godot.windows.tools.64.exe!TestObject::_DOCTEST_ANON_FUNC_1791() Line 219 (tests\test_object.h:219)
> godot.windows.tools.64.exe!doctest::Context::run() Line 6168 (thirdparty\doctest\doctest.h:6168)
> godot.windows.tools.64.exe!test_main(int argc, char * * argv) Line 130 (tests\test_main.cpp:130)
> godot.windows.tools.64.exe!Main::test_entrypoint(int argc, char * * argv, bool & tests_need_run) Line 469 (main\main.cpp:469)
> godot.windows.tools.64.exe!widechar_main(int argc, wchar_t * * argv) Line 149 (platform\windows\godot_windows.cpp:149)
> godot.windows.tools.64.exe!_main() Line 185 (platform\windows\godot_windows.cpp:185)
> godot.windows.tools.64.exe!main(int argc, char * * argv) Line 199 (platform\windows\godot_windows.cpp:199)
> [Inline Frame] godot.windows.tools.64.exe!invoke_main() Line 78 (d:\agent\_work\63\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78)
> godot.windows.tools.64.exe!__scrt_common_main_seh() Line 288 (d:\agent\_work\63\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
> kernel32.dll!00007ffb60027c24() (Unknown Source:0)
> ntdll.dll!00007ffb6022d4d1() (Unknown Source:0)
> ```
> 
> You'll also want to run the binary with `--test --no-breaks` so the debugger won't break on ClassDB tests currently.

Interesting, I am also running the test on Windows (project compiled with MSVC) and it finishes successfully.

File: tests/test_object.h [hunk 3]
```diff
@@ -87,6 +187,119 @@ TEST_CASE("[Object] Metadata") {
...
+	bool valid = false;
+	const Variant &actual_value = object.get("some_name", &valid);
+	CHECK(valid);
+	CHECK_MESSAGE(
+			actual_value == Variant(100),
+			"The returned value should equal the one which was set by the script instance.");
+}
+
+TEST_CASE("[Object] Built-in property setter") {
+	ClassDB::register_class<_TestDerivedObject>();
```
Comment (Xrayez) [inline tests/test_object.h:236]: Just wanted to note that `ClassDB` is used globally throughout tests, but as long as it doesn't affect other tests, this should be fine I think.

**Comment by Xrayez** [comment]
> Interesting, I am running the test on Windows 10 (project compiled with MSVC) and it finishes successfully.

It depends on how you run tests, a Godot project may certainly display a backtrace upon crash (given debug symbols are enabled). If I run tests without debugger, then I see no apparent crash either.

**Review by Xrayez — APPROVED** [review]
Should be good to merge!

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**